File: frontend/lib/models/boutique.dart
import '../core/url_utils.dart';

class Boutique {
  final String id;
  final String name;
  final String slug;
  final String? logoUrl;
  final String? description;
  final String? email;
  final String? phone;
  final String? address;
  final String? currency;
  final String? language;
  final String? timezone;
  final bool isActive;
  final String? customDomain;
  final String? primaryColor;
  final String? secondaryColor;
  final String? seoTitle;
  final String? seoDescription;
  final String? seoKeywords;
  final String? ogImageUrl;
  final String? facebookUrl;
  final String? instagramUrl;
  final String? tiktokUrl;
  final String? twitterUrl;
  final String? linkedinUrl;
  final String? whatsappNumber;
  final bool enableCod;
  final bool enableJax;
  final bool enableIntigo;
  final bool enableAdeex;

  // Storefront config
  final String? storeConfig;
  final String? headerColor;
  final String? footerColor;
  final String? bodyColor;
  final String? cardProductColor;
  final String? buttonColor;
  final String? topBarColor;
  final String? textColor;
  final String? announcementText;
  final double? deliveryFees;
  final double? tva;
  final bool simpleCheckout;
  final bool cashOnDelivery;
  final String? facebookPixelId;
  final String? googleAnalyticsId;

  // New store settings
  final String? bannerUrl;
  final String? faviconUrl;
  final String? fontFamily;
  final bool darkMode;
  final bool stripeEnabled;
  final String? stripeStatus;
  final String? stripePublishableKey;
  final double? freeShippingThreshold;
  final int? estimatedDeliveryDays;
  final bool enableLocalPickup;
  final bool enableEmailNotifications;
  final bool enableSmsNotifications;
  final bool enablePushNotifications;
  final bool enableMarketingEmails;
  final bool enableOrderAlerts;
  final bool clientMessagingEnabled;
  final bool teamEnabled;
  final String? telegramChatId;
  final bool telegramEnabled;

  final String? storeStatus;
  final String? publicationStatus;
  final String? frozenAt;
  final String? freezeReason;
  final bool isPublished;
  final String? publishedAt;
  final String? publicUrl;
  final bool ownerAccess;
  final String? currentUserRole;
  final List<String> currentUserPermissions;

  Boutique({
    required this.id,
    required this.name,
    required this.slug,
    this.logoUrl,
    this.description,
    this.email,
    this.phone,
    this.address,
    this.currency,
    this.language,
    this.timezone,
    required this.isActive,
    this.customDomain,
    this.primaryColor,
    this.secondaryColor,
    this.seoTitle,
    this.seoDescription,
    this.seoKeywords,
    this.ogImageUrl,
    this.facebookUrl,
    this.instagramUrl,
    this.tiktokUrl,
    this.twitterUrl,
    this.linkedinUrl,
    this.whatsappNumber,
    this.enableCod = true,
    this.enableJax = false,
    this.enableIntigo = false,
    this.enableAdeex = false,
    this.storeConfig,
    this.headerColor,
    this.footerColor,
    this.bodyColor,
    this.cardProductColor,
    this.buttonColor,
    this.topBarColor,
    this.textColor,
    this.announcementText,
    this.deliveryFees,
    this.tva,
    this.simpleCheckout = false,
    this.cashOnDelivery = true,
    this.facebookPixelId,
    this.googleAnalyticsId,
    this.bannerUrl,
    this.faviconUrl,
    this.fontFamily,
    this.darkMode = false,
    this.stripeEnabled = false,
    this.stripeStatus,
    this.stripePublishableKey,
    this.freeShippingThreshold,
    this.estimatedDeliveryDays,
    this.enableLocalPickup = false,
    this.enableEmailNotifications = true,
    this.enableSmsNotifications = false,
    this.enablePushNotifications = true,
    this.enableMarketingEmails = false,
    this.enableOrderAlerts = true,
    this.clientMessagingEnabled = true,
    this.teamEnabled = false,
    this.telegramChatId,
    this.telegramEnabled = false,
    this.storeStatus,
    this.publicationStatus,
    this.frozenAt,
    this.freezeReason,
    this.isPublished = false,
    this.publishedAt,
    this.publicUrl,
    this.ownerAccess = false,
    this.currentUserRole,
    this.currentUserPermissions = const [],
  });

  factory Boutique.fromJson(Map<String, dynamic> json) => Boutique(
    id: json['id'].toString(),
    name: json['name'] ?? '',
    slug: json['slug'] ?? '',
    logoUrl: normalizeRemoteUrl(json['logoUrl']),
    description: json['description'],
    email: json['email'],
    phone: json['phone'],
    address: json['address'],
    currency: json['currency'],
    language: json['language'],
    timezone: json['timezone'],
    isActive: json['isActive'] ?? true,
    customDomain: json['customDomain'],
    primaryColor: json['primaryColor'],
    secondaryColor: json['secondaryColor'],
    seoTitle: json['seoTitle'],
    seoDescription: json['seoDescription'],
    seoKeywords: json['seoKeywords'],
    ogImageUrl: normalizeRemoteUrl(json['ogImageUrl']),
    facebookUrl: json['facebookUrl'],
    instagramUrl: json['instagramUrl'],
    tiktokUrl: json['tiktokUrl'],
    twitterUrl: json['twitterUrl'],
    linkedinUrl: json['linkedinUrl'],
    whatsappNumber: json['whatsappNumber'],
    enableCod: json['enableCod'] ?? true,
    enableJax: json['enableJax'] ?? false,
    enableIntigo: json['enableIntigo'] ?? false,
    enableAdeex: json['enableAdeex'] ?? false,
    storeConfig: json['storeConfig'],
    headerColor: json['headerColor'] ?? json['header_color'],
    footerColor: json['footerColor'] ?? json['footer_color'],
    bodyColor: json['bodyColor'] ?? json['body_color'],
    cardProductColor: json['cardProductColor'] ?? json['card_product_color'],
    buttonColor: json['buttonColor'] ?? json['button_color'],
    topBarColor: json['topBarColor'] ?? json['top_bar_color'],
    textColor: json['textColor'] ?? json['text_color'],
    announcementText: json['announcementText'] ?? json['announcement_text'],
    deliveryFees: (json['deliveryFees'] ?? json['delivery_fees'] ?? 7.0).toDouble(),
    tva: (json['tva'] ?? 0.0).toDouble(),
    simpleCheckout: json['simpleCheckout'] ?? json['simple_checkout'] ?? false,
    cashOnDelivery: json['cashOnDelivery'] ?? json['cash_on_delivery'] ?? true,
    facebookPixelId: json['facebookPixelId'] ?? json['facebook_pixel_id'],
    googleAnalyticsId: json['googleAnalyticsId'] ?? json['google_analytics_id'],
    bannerUrl: normalizeRemoteUrl(json['bannerUrl']),
    faviconUrl: json['faviconUrl'],
    fontFamily: json['fontFamily'],
    darkMode: json['darkMode'] ?? false,
    stripeEnabled: json['stripeEnabled'] ?? ((json['stripeStatus'] ?? json['stripe_status'])?.toString().toLowerCase() == 'active' || (json['stripeStatus'] ?? json['stripe_status'])?.toString().toLowerCase() == 'enabled'),
    stripeStatus: json['stripeStatus'] ?? json['stripe_status'],
    stripePublishableKey: json['stripePublishableKey'],
    freeShippingThreshold: (json['freeShippingThreshold'] as num?)?.toDouble(),
    estimatedDeliveryDays: json['estimatedDeliveryDays'],
    enableLocalPickup: json['enableLocalPickup'] ?? false,
    enableEmailNotifications: json['enableEmailNotifications'] ?? true,
    enableSmsNotifications: json['enableSmsNotifications'] ?? false,
    enablePushNotifications: json['enablePushNotifications'] ?? true,
    enableMarketingEmails: json['enableMarketingEmails'] ?? false,
    enableOrderAlerts: json['enableOrderAlerts'] ?? true,
    clientMessagingEnabled: json['clientMessagingEnabled'] ?? true,
    teamEnabled: json['teamEnabled'] ?? false,
    telegramChatId: json['telegramChatId'],
    telegramEnabled: json['telegramEnabled'] ?? false,
    storeStatus: json['storeStatus'],
    publicationStatus: json['publicationStatus'],
    frozenAt: json['frozenAt'],
    freezeReason: json['freezeReason'],
    isPublished: json['isPublished'] ?? false,
    publishedAt: json['publishedAt'],
    publicUrl: json['publicUrl'],
    ownerAccess: json['ownerAccess'] == true,
    currentUserRole: json['currentUserRole'],
    currentUserPermissions: (json['currentUserPermissions'] as List<dynamic>?)
        ?.map((e) => e.toString())
        .toList() ?? const [],
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'slug': slug,
    'logoUrl': logoUrl,
    'description': description,
    'email': email,
    'phone': phone,
    'address': address,
    'currency': currency,
    'language': language,
    'timezone': timezone,
    'isActive': isActive,
    'customDomain': customDomain,
    'primaryColor': primaryColor,
    'secondaryColor': secondaryColor,
    'seoTitle': seoTitle,
    'seoDescription': seoDescription,
    'seoKeywords': seoKeywords,
    'ogImageUrl': ogImageUrl,
    'facebookUrl': facebookUrl,
    'instagramUrl': instagramUrl,
    'tiktokUrl': tiktokUrl,
    'twitterUrl': twitterUrl,
    'linkedinUrl': linkedinUrl,
    'whatsappNumber': whatsappNumber,
    'enableCod': enableCod,
    'enableJax': enableJax,
    'enableIntigo': enableIntigo,
    'enableAdeex': enableAdeex,
    'headerColor': headerColor,
    'footerColor': footerColor,
    'bodyColor': bodyColor,
    'cardProductColor': cardProductColor,
    'buttonColor': buttonColor,
    'topBarColor': topBarColor,
    'textColor': textColor,
    'announcementText': announcementText,
    'deliveryFees': deliveryFees,
    'tva': tva,
    'simpleCheckout': simpleCheckout,
    'cashOnDelivery': cashOnDelivery,
    'facebookPixelId': facebookPixelId,
    'googleAnalyticsId': googleAnalyticsId,
    'bannerUrl': bannerUrl,
    'faviconUrl': faviconUrl,
    'fontFamily': fontFamily,
    'darkMode': darkMode,
    'stripeEnabled': stripeEnabled,
    'stripeStatus': stripeStatus,
    'stripePublishableKey': stripePublishableKey,
    'freeShippingThreshold': freeShippingThreshold,
    'estimatedDeliveryDays': estimatedDeliveryDays,
    'enableLocalPickup': enableLocalPickup,
    'enableEmailNotifications': enableEmailNotifications,
    'enableSmsNotifications': enableSmsNotifications,
    'enablePushNotifications': enablePushNotifications,
    'enableMarketingEmails': enableMarketingEmails,
    'enableOrderAlerts': enableOrderAlerts,
    'clientMessagingEnabled': clientMessagingEnabled,
    'teamEnabled': teamEnabled,
    'telegramChatId': telegramChatId,
    'telegramEnabled': telegramEnabled,
    'storeStatus': storeStatus,
    'publicationStatus': publicationStatus,
    'frozenAt': frozenAt,
    'freezeReason': freezeReason,
    'isPublished': isPublished,
    'publishedAt': publishedAt,
    'publicUrl': publicUrl,
    'ownerAccess': ownerAccess,
    'currentUserRole': currentUserRole,
    'currentUserPermissions': currentUserPermissions,
  };

  bool hasPermission(String permission) =>
      ownerAccess || currentUserPermissions.contains(permission);

  bool hasAnyPermission(List<String> permissions) =>
      ownerAccess || permissions.any(currentUserPermissions.contains);
}

class BoutiqueStats {
  final int totalOrders;
  final int todayOrders;
  final double totalRevenue;
  final double todayRevenue;
  final int totalProducts;
  final int pendingOrders;
  final String? activeBoutiqueId;

  BoutiqueStats({
    required this.totalOrders,
    required this.todayOrders,
    required this.totalRevenue,
    required this.todayRevenue,
    required this.totalProducts,
    required this.pendingOrders,
    this.activeBoutiqueId,
  });

  factory BoutiqueStats.fromJson(Map<String, dynamic> json) => BoutiqueStats(
    totalOrders: json['totalOrders'] ?? 0,
    todayOrders: json['todayOrders'] ?? 0,
    totalRevenue: (json['totalRevenue'] ?? 0).toDouble(),
    todayRevenue: (json['todayRevenue'] ?? 0).toDouble(),
    totalProducts: json['totalProducts'] ?? 0,
    pendingOrders: json['pendingOrders'] ?? 0,
    activeBoutiqueId: json['activeBoutiqueId'],
  );
}

File: backend/src/main/java/io/makewebsite/entity/Boutique.java
package io.makewebsite.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.GenericGenerator;

import java.time.LocalDateTime;
import java.util.UUID;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "boutiques")
public class Boutique {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(updatable = false)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "tenant_id", nullable = false)
    private Tenant tenant;

    @NotNull
    @Size(max = 100)
    @Column(nullable = false)
    private String name;

    @NotNull
    @Size(max = 100)
    @Column(unique = true, nullable = false)
    private String slug;

    @Column(name = "logo_url", columnDefinition = "TEXT")
    private String logoUrl;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "custom_domain", columnDefinition = "TEXT")
    private String customDomain;

    @Builder.Default
    @Size(max = 7)
    @Column(name = "primary_color")
    private String primaryColor = "#2710BF";

    @Builder.Default
    @Size(max = 7)
    @Column(name = "secondary_color")
    private String secondaryColor = "#6C4FFF";

    @Builder.Default
    @Size(max = 10)
    private String currency = "TND";

    @Builder.Default
    @Size(max = 10)
    private String language = "fr";

    @Size(max = 100)
    private String category;

    @Size(max = 100)
    private String country;

    @Size(max = 100)
    private String city;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @Size(max = 200)
    @Column(name = "seo_title")
    private String seoTitle;

    @Column(name = "seo_description", columnDefinition = "TEXT")
    private String seoDescription;

    @Column(name = "seo_keywords", columnDefinition = "TEXT")
    private String seoKeywords;

    @Column(name = "facebook_url", columnDefinition = "TEXT")
    private String facebookUrl;

    @Column(name = "instagram_url", columnDefinition = "TEXT")
    private String instagramUrl;

    @Column(name = "tiktok_url", columnDefinition = "TEXT")
    private String tiktokUrl;

    @Column(name = "twitter_url", columnDefinition = "TEXT")
    private String twitterUrl;

    @Column(name = "linkedin_url", columnDefinition = "TEXT")
    private String linkedinUrl;

    @Size(max = 20)
    @Column(name = "whatsapp_number")
    private String whatsappNumber;

    @Column(name = "custom_css", columnDefinition = "TEXT")
    private String customCss;

    @Column(name = "custom_js", columnDefinition = "TEXT")
    private String customJs;

    @Builder.Default
    @Column(name = "enable_cod")
    private Boolean enableCod = true;

    @Builder.Default
    @Column(name = "enable_d17")
    private Boolean enableD17 = false;

    @Builder.Default
    @Column(name = "enable_adeex")
    private Boolean enableAdeex = false;

    @Builder.Default
    @Column(name = "enable_jax")
    private Boolean enableJax = false;

    @Builder.Default
    @Column(name = "enable_intigo")
    private Boolean enableIntigo = false;

    // ====== Storefront config fields ======

    @Builder.Default
    @Column(name = "store_config", columnDefinition = "TEXT")
    private String storeConfig = "{}";

    @Column(name = "generated_html", columnDefinition = "TEXT")
    private String generatedHtml;

    @Builder.Default
    @Column(name = "header_color", length = 7)
    private String headerColor = "#ededed";

    @Builder.Default
    @Column(name = "footer_color", length = 7)
    private String footerColor = "#dbdbdb";

    @Builder.Default
    @Column(name = "body_color", length = 7)
    private String bodyColor = "#ffffff";

    @Builder.Default
    @Column(name = "card_product_color", length = 7)
    private String cardProductColor = "#fafafa";

    @Builder.Default
    @Column(name = "button_color", length = 7)
    private String buttonColor = "#b551c2";

    @Builder.Default
    @Column(name = "top_bar_color", length = 7)
    private String topBarColor = "#3b0086";

    @Builder.Default
    @Column(name = "text_color", length = 7)
    private String textColor = "#751515";

    @Column(name = "announcement_text", columnDefinition = "TEXT")
    private String announcementText;

    @Builder.Default
    @Column(name = "delivery_fees")
    private Double deliveryFees = 7.00;

    @Builder.Default
    @Column(name = "tva")
    private Double tva = 0.00;

    @Builder.Default
    @Column(name = "simple_checkout")
    private Boolean simpleCheckout = false;

    @Builder.Default
    @Column(name = "cash_on_delivery")
    private Boolean cashOnDelivery = true;

    @Column(name = "konnect_merchant_id", length = 100)
    private String konnectMerchantId;

    @Column(name = "konnect_api_key", length = 200)
    private String konnectApiKey;

    @Builder.Default
    @Column(name = "konnect_status", length = 20)
    private String konnectStatus = "inactive";

    @Column(name = "d17_merchant_number", length = 50)
    private String d17MerchantNumber;

    @Column(name = "d17_qr_code_url", columnDefinition = "TEXT")
    private String d17QrCodeUrl;

    @Builder.Default
    @Column(name = "d17_status", length = 20)
    private String d17Status = "inactive";

    @Builder.Default
    @Column(name = "invoice_sequence")
    private Long invoiceSequence = 0L;

    @Column(name = "facebook_pixel_id", length = 50)
    private String facebookPixelId;

    @Column(name = "google_analytics_id", length = 50)
    private String googleAnalyticsId;

    // ====== New store settings fields ======

    @Column(length = 200)
    private String email;

    @Column(length = 50)
    private String phone;

    @Column(columnDefinition = "TEXT")
    private String address;

    @Builder.Default
    @Column(length = 50)
    private String timezone = "Africa/Tunis";

    @Column(name = "banner_url", columnDefinition = "TEXT")
    private String bannerUrl;

    @Column(name = "favicon_url", columnDefinition = "TEXT")
    private String faviconUrl;

    @Column(name = "og_image_url", columnDefinition = "TEXT")
    private String ogImageUrl;

    @Builder.Default
    @Column(name = "font_family", length = 100)
    private String fontFamily = "Inter";

    @Builder.Default
    @Column(name = "dark_mode")
    private Boolean darkMode = false;

    @Column(name = "stripe_publishable_key", length = 200)
    private String stripePublishableKey;

    @Column(name = "stripe_secret_key", length = 200)
    private String stripeSecretKey;

    @Column(name = "stripe_webhook_secret", length = 200)
    private String stripeWebhookSecret;

    @Column(name = "free_shipping_threshold")
    private Double freeShippingThreshold;

    @Builder.Default
    @Column(name = "estimated_delivery_days")
    private Integer estimatedDeliveryDays = 3;

    @Builder.Default
    @Column(name = "enable_local_pickup")
    private Boolean enableLocalPickup = false;

    @Builder.Default
    @Column(name = "enable_email_notifications")
    private Boolean enableEmailNotifications = true;

    @Builder.Default
    @Column(name = "enable_sms_notifications")
    private Boolean enableSmsNotifications = false;

    @Builder.Default
    @Column(name = "enable_push_notifications")
    private Boolean enablePushNotifications = true;

    @Builder.Default
    @Column(name = "enable_marketing_emails")
    private Boolean enableMarketingEmails = false;

    @Builder.Default
    @Column(name = "enable_order_alerts")
    private Boolean enableOrderAlerts = true;

    @Column(name = "template_id")
    private Integer templateId;

    @Builder.Default
    @Column(name = "client_messaging_enabled")
    private Boolean clientMessagingEnabled = true;

    @Builder.Default
    @Column(name = "team_enabled")
    private Boolean teamEnabled = false;

    @Builder.Default
    @Column(name = "store_status", length = 20)
    private String storeStatus = "ACTIVE";

    @Column(name = "frozen_at")
    private LocalDateTime frozenAt;

    @Column(name = "freeze_reason", columnDefinition = "TEXT")
    private String freezeReason;

    @Builder.Default
    @Column(name = "is_published")
    private Boolean isPublished = false;

    @Column(name = "published_at")
    private LocalDateTime publishedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}
